src/common/column_matrix.h

```diff
@@ -366,19 +423,107 @@ class ColumnMatrix {
       using ColumnBinT = decltype(t);
       ColumnBinT* local_index = reinterpret_cast<ColumnBinT*>(index_.data());
       size_t const batch_size = batch.Size();
-      size_t k{0};
-      for (size_t rid = 0; rid < batch_size; ++rid) {
-        auto line = batch.GetLine(rid);
-        for (size_t i = 0; i < line.Size(); ++i) {
-          auto coo = line.GetElement(i);
-          if (is_valid(coo)) {
-            auto fid = coo.column_idx;
-            const uint32_t bin_id = row_index[k];
-            SetBinSparse(bin_id, rid + base_rowid, fid, local_index);
-            ++k;
+
+      // Parallel sparse batch processing
+      dmlc::OMPException exc;
+      std::vector<size_t> n_elements((n_threads + 1) * n_features, 0);
+      std::vector<size_t> k_offsets(n_threads + 1, 0);
```

Comment:
Description of the purpose of those arrays:
```suggestion
            std::vector<size_t> n_elements((n_threads + 1) * n_features, 0);  // number of non-zero elements per feature per thread
                                                                        // n_elements[tid * n_features + fid] =
                                                                        //   number of non-zero elements of feature fid processed by thread tid
                                                                        // n_elements[n_threads * n_features + fid] =
                                                                        //   total number of non-zero elements of feature fid
      std::vector<size_t> k_offsets(n_threads + 1, 0);        // offsets of non-zero elements for each thread
                                                              // k_offsets[0] = 0;
                                                              // k_offsets[tid] - starting offset of non-zero elements processed by thread tid
```